{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SimiC Preprocessing Pipeline Tutorial\n",
    "\n",
    ">*Author: Irene Marín-Goñi, PhD student - ML4BM group (CIMA University of Navarra)*\n",
    "\n",
    "This notebook demonstrates how to preprocess single-cell RNA-seq data for SimiC analysis.\n",
    "\n",
    "## Overview\n",
    "This preprocessing tutorial covers:\n",
    "1. Package installation and setup\n",
    "2. MAGIC imputation pipeline\n",
    "3. Gene selection and experiment setup\n",
    "4. Preparing input files for SimiC\n",
    "\n",
    "For running SimiC analysis see `Tutorial_SimiCPipeline_simple.ipynb`\n",
    "\n",
    "## Introduction\n",
    "Before running SimiC, you need to:\n",
    "- Impute your scRNA-seq data. We recommend to use [MAGIC](https://pypi.org/project/magic-impute/3.0.0/) and include a wrapper class `MagicPipeline` to ease the process.\n",
    "- Select top variable genes based on MAD (Median Absolute Deviation).\n",
    "- Prepare input files in the correct format.\n",
    "\n",
    "This tutorial shows you how to do all of this using the SimiC preprocessing modules."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup\n",
    "\n",
    "<div class=\"alert alert-block alert-warning\">\n",
    "<b>Warning: Need to include here Installing instructions (github/Docker/dependencies)</b>\n",
    "</div>\n",
    "\n",
    "Required packages:\n",
    "- anndata\n",
    "- pandas\n",
    "- numpy\n",
    "- scipy\n",
    "- scprep\n",
    "- magic-impute"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import Modules\n",
    "First, import the necessary preprocessing modules."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/md0/imaring/SimiCPipeline\n",
      "/mnt/md0/imaring/test\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "print(os.getcwd())\n",
    "os.chdir('../test')\n",
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<built-in function getcwd>\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'/mnt/md0/imaring/SimiCPipeline'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mCanceled future for execute_request message before replies were done"
     ]
    },
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mFailed to connect to the remote Jupyter Server 'http://127.0.0.1:8889/'. Verify the server is running and reachable."
     ]
    },
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mCanceled future for execute_request message before replies were done. \n",
      "\u001b[1;31mView Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details."
     ]
    }
   ],
   "source": [
    "import os\n",
    "print(os.getcwd)\n",
    "os.chdir('/home/workdir')\n",
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0.1.0'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import simicpipeline \n",
    "simicpipeline.__version__"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part 1: MAGIC Imputation Pipeline\n",
    "\n",
    "MAGIC (Markov Affinity-based Graph Imputation of Cells) is used to denoise and impute scRNA-seq data. `MagicPipeline` facilitates the steps described in [Magic Tutorial](\"https://magic.readthedocs.io/en/stable/tutorial.html\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Step 1.1: Load Your Data\n",
    "\n",
    "Load your raw expression data. Your input should be in AnnData format. \n",
    "You can see an example below starting from pandas DataFrame or 10x format.\n",
    "\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "<em> <b>Note:</b> If you have already processsed your data, make sure the adata object has the raw counts in the `adata.raw.X` slot</b>\n",
    "</div>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import pandas as pd\n",
    "# import anndata as ad\n",
    "# Example: Load from h5ad file\n",
    "# adata = ad.read_h5ad('path/to/your/data.h5ad')\n",
    "\n",
    "# Example: Load from 10X format\n",
    "# adata = ad.read_10x_mtx('path/to/10x/directory')\n",
    "\n",
    "# Example: Load from CSV files\n",
    "# expression_data = pd.read_csv('path/to/expression_data.csv', index_col=0)\n",
    "# metadata = pd.read_csv('path/to/metadata.csv', index_col=0)\n",
    "# adata = ad.AnnData(X=expression_data.values, obs=metadata)\n",
    "\n",
    "# Example: Load from Matrix Market format\n",
    "\n",
    "# df = simicpipeline.load_from_matrix_market( \n",
    "#     matrix_path=Path(\"path/to/matrix.mtx\"),\n",
    "#     genes_path=Path(\"path/to/genes.txt\"),\n",
    "#     cells_path=Path(\"path/to/cells.txt\"),\n",
    "#     transpose=False,\n",
    "#     cells_index_name=\"Cell\",\n",
    "# )\n",
    "# adata = ad.AnnData(X=df.values, obs=pd.DataFrame(index=df.index), var=pd.DataFrame(index=df.columns))\n",
    "# If your data is raw, you should set it properly \n",
    "# adata.raw = adata.copy()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Load your AnnData object here\n",
      "<class 'scipy.sparse._csr.csr_matrix'>\n",
      "<class 'anndata._core.anndata.AnnData'>\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 72650 × 36774\n",
       "    obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'sample', 'species', 'gene_count', 'tscp_count', 'mread_count', 'bc1_wind', 'bc2_wind', 'bc3_wind', 'bc1_well', 'bc2_well', 'bc3_well', 'treatment', 'cellLine', 'percent.mt', 'percent.rb', 'percent.hb', 'integrated_snn_res.0.095', 'seurat_clusters', 'sctype_custom', 'sctype_functional', 'sctype_SCsubtype', 'final_annotation', 'final_annotation_functional'\n",
       "    var: 'vf_vst_counts.1_mean', 'vf_vst_counts.1_variance', 'vf_vst_counts.1_variance.expected', 'vf_vst_counts.1_variance.standardized', 'vf_vst_counts.1_variable', 'vf_vst_counts.1_rank', 'vf_vst_counts.2_mean', 'vf_vst_counts.2_variance', 'vf_vst_counts.2_variance.expected', 'vf_vst_counts.2_variance.standardized', 'vf_vst_counts.2_variable', 'vf_vst_counts.2_rank', 'vf_vst_counts.3_mean', 'vf_vst_counts.3_variance', 'vf_vst_counts.3_variance.expected', 'vf_vst_counts.3_variance.standardized', 'vf_vst_counts.3_variable', 'vf_vst_counts.3_rank', 'vf_vst_counts.4_mean', 'vf_vst_counts.4_variance', 'vf_vst_counts.4_variance.expected', 'vf_vst_counts.4_variance.standardized', 'vf_vst_counts.4_variable', 'vf_vst_counts.4_rank', 'vf_vst_counts.5_mean', 'vf_vst_counts.5_variance', 'vf_vst_counts.5_variance.expected', 'vf_vst_counts.5_variance.standardized', 'vf_vst_counts.5_variable', 'vf_vst_counts.5_rank', 'vf_vst_counts.6_mean', 'vf_vst_counts.6_variance', 'vf_vst_counts.6_variance.expected', 'vf_vst_counts.6_variance.standardized', 'vf_vst_counts.6_variable', 'vf_vst_counts.6_rank', 'vf_vst_counts.7_mean', 'vf_vst_counts.7_variance', 'vf_vst_counts.7_variance.expected', 'vf_vst_counts.7_variance.standardized', 'vf_vst_counts.7_variable', 'vf_vst_counts.7_rank', 'vf_vst_counts.8_mean', 'vf_vst_counts.8_variance', 'vf_vst_counts.8_variance.expected', 'vf_vst_counts.8_variance.standardized', 'vf_vst_counts.8_variable', 'vf_vst_counts.8_rank', 'vf_vst_counts_mean', 'vf_vst_counts_variance', 'vf_vst_counts_variance.expected', 'vf_vst_counts_variance.standardized', 'vf_vst_counts_variable', 'vf_vst_counts_rank', 'var.features', 'var.features.rank'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import anndata as ad\n",
    "print(\"Load your AnnData object here\")\n",
    "# from pathlib import Path\n",
    "# df = simicpipeline.load_from_matrix_market( \n",
    "#     matrix_path=Path(\"./data/simic_matrix.txt\"),\n",
    "#     genes_path=Path(\"./data/simic_genes.txt\"),\n",
    "#     cells_path=Path(\"./data/simic_cells.txt\"),\n",
    "#     transpose=True,\n",
    "#     cells_index_name=\"Cell\",\n",
    "# )\n",
    "# adata = ad.AnnData(X=df.values, obs=pd.DataFrame(index=df.index), var=pd.DataFrame(index=df.columns))\n",
    "# adata.raw = adata.copy()\n",
    "# adata = simicpipeline.load_from_anndata('path/to/your/data.h5ad')\n",
    "adata = simicpipeline.load_from_anndata('./data/DAC_aPDL1_seurat_annotated.h5ad')\n",
    "print(type(adata.X))\n",
    "print(type(adata))\n",
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1.2: Initialize MAGIC Pipeline\n",
    "\n",
    "Create a MAGIC pipeline instance:\n",
    "- `input_data`: Your AnnData object. If you run the full pipline starting from raw counts they should be in `adata.X.raw`\n",
    "- `project_dir`: Project directory where `magic_output`dir will be created and files will be saved\n",
    "- `magic_output_file`: Filename for the imputed data (default: 'magic_data_allcells_sqrt.pickle')\n",
    "- `filtered`: Set to True if data is already filtered (low quality cells and genes) (default: False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MagicPipeline(\n",
      "  data = AnnData object with (n_obs × n_vars) = 72650 × 36774,\n",
      "  filtered = False,\n",
      "  imputed = False,\n",
      "  magic_data = None,\n",
      "  project_dir = 'SimiCExampleRun'\n",
      ")\n"
     ]
    }
   ],
   "source": [
    "# This command will initialize the MAGIC pipeline and generate the output directory if it does not exist\n",
    "from simicpipeline import MagicPipeline\n",
    "magic_pipeline = MagicPipeline(\n",
    "    input_data= adata,\n",
    "    project_dir='./SimiCExampleRun',\n",
    "    magic_output_file='magic_imputed.pickle',\n",
    "    filtered=False\n",
    ")\n",
    "\n",
    "print(magic_pipeline)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1.3: Filter Cells and Genes\n",
    "\n",
    "Remove low-quality cells and lowly-expressed genes:\n",
    "- `min_cells_per_gene`: Minimum number of cells expressing a gene (default: 10)\n",
    "- `min_umis_per_cell`: Minimum total UMI counts per cell (default: 500)\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "<em><b>Note:</b> If your data was already filtered you can skip this step and set the flitered argument flag to `True` in the previous step.</em>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Filtering cells and genes...\n",
      "Before filtering: 72650 cells x 36774 genes\n",
      "Keeping 27837/36774 genes (75.70%)\n",
      "Keeping 72650/72650 cells (100.00%)\n",
      "All cells pass the filter!\n",
      "After filtering: 72650 cells x 27837 genes\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "MagicPipeline(\n",
       "  data = AnnData object with (n_obs × n_vars) = 72650 × 27837,\n",
       "  filtered = True,\n",
       "  imputed = False,\n",
       "  magic_data = None,\n",
       "  project_dir = 'SimiCExampleRun'\n",
       ")"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "magic_pipeline.filter_cells_and_genes(min_cells_per_gene = 10, min_umis_per_cell = 500)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1.4: Normalize Data\n",
    "\n",
    "Perform library size normalization with `scprep` followed by square root transformation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Normalizing data...\n",
      "After normalization: 72650 cells x 27837 genes\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "MagicPipeline(\n",
       "  data = AnnData object with (n_obs × n_vars) = 72650 × 27837,\n",
       "  filtered = True,\n",
       "  imputed = False,\n",
       "  magic_data = None,\n",
       "  project_dir = 'SimiCExampleRun'\n",
       ")"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "magic_pipeline.normalize_data()\n",
    "# Note this will overide adata.X with normalized data and remove adata.raw slot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1.5: Run MAGIC Imputation\n",
    "\n",
    "Run MAGIC imputation with custom parameters:\n",
    "- `t`: Number of diffusion steps (default: 'auto')\n",
    "- `knn`: Number of nearest neighbors (default: 5)\n",
    "- `decay`: Decay rate for kernel (default: 1)\n",
    "- `n_jobs`: Number of parallel jobs (default: -2)\n",
    "- `genes`: Genes to be returned. If None or \"all genes\" it returns teh entire matrix.\n",
    "- `save_data`: Whether to automatically save imputed data (default: True). If magic_output_file extension is .pickle will save it in .pickle, if h5ad, will save in adata format.\n",
    "\n",
    "See [MAGIC documentation](https://magic.readthedocs.io/) for more parameter options."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Running MAGIC imputation...\n",
      "Calculating MAGIC...\n",
      "  Running MAGIC on 72650 cells and 27837 genes.\n",
      "  Calculating graph and diffusion operator...\n",
      "    Calculating PCA...\n",
      "    Calculated PCA in 193.51 seconds.\n",
      "    Calculating KNN search...\n",
      "    Calculated KNN search in 3.19 seconds.\n",
      "    Calculating affinities...\n",
      "    Calculated affinities in 7.04 seconds.\n",
      "  Calculated graph and diffusion operator in 203.88 seconds.\n",
      "  Running MAGIC with `solver='exact'` on 27837-dimensional data may take a long time. Consider denoising specific genes with `genes=<list-like>` or using `solver='approximate'`.\n",
      "  Calculating imputation...\n",
      "  Calculated imputation in 160.10 seconds.\n",
      "Calculated MAGIC in 377.39 seconds.\n",
      "MAGIC imputation complete:  72650 cells x 27837 genes\n",
      "\n",
      "Saving MAGIC-imputed data to SimiCExampleRun/magic_output/magic_imputed.pickle...\n",
      "Saved successfully to SimiCExampleRun/magic_output/magic_imputed.pickle\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "MagicPipeline(\n",
       "  data = AnnData object with (n_obs × n_vars) = 72650 × 27837,\n",
       "  filtered = True,\n",
       "  imputed = True,\n",
       "  magic_data = AnnData object with n_obs × n_vars = 72650 × 27837,\n",
       "  project_dir = 'SimiCExampleRun'\n",
       ")"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "magic_pipeline.run_magic(\n",
    "    random_state=123,\n",
    "    n_jobs=-2,  # Use all but 1 CPU cores\n",
    "    save_data=True\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1.6: Check Pipeline Status"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "MagicPipeline(\n",
       "  data = AnnData object with (n_obs × n_vars) = 72650 × 27837,\n",
       "  filtered = True,\n",
       "  imputed = True,\n",
       "  magic_data = AnnData object with n_obs × n_vars = 72650 × 27837,\n",
       "  project_dir = 'SimiCExampleRun'\n",
       ")"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "magic_pipeline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "<b>Success!</b> MAGIC imputation is complete. The imputed data is saved in the magic_output directory.\n",
    "</div>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "# Part 2: Experiment Setup and Gene Selection\n",
    "\n",
    "Now we will select top variable genes and prepare input files for SimiCPipeline. \n",
    "<div class=\"alert alert-block alert-info\">\n",
    "\n",
    "<em><b>Note:</b> if you have your data filtered, normalized and inputed with other methods you can start from here.</em>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Step 2.1: Load Imputed Data\n",
    "\n",
    "In this example we will start from the imputed AnnData object from the MAGIC pipeline. If you saved and stopped your work, you can re-load the object with the following code:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# If you saved it in h5ad format, you can load it back using:\n",
    "# import simicpipeline\n",
    "# imputed_data = simicpipeline.load_from_anndata('path/to/magic_imputed.h5ad')\n",
    "# # If you saved it in pickle format, you can load it back using:\n",
    "# import pickle\n",
    "# with open('path/to/magic_imputed.pickle', 'rb') as f:\n",
    "#     imputed_data = pickle.load(f)\n",
    "# If you contintue from the previous section you can access the MAGIC-imputed AnnData object\n",
    "# imputed_data = magic_pipeline.magic_adata\n",
    "\n",
    "# If you have inputed data from other methods, you can directly use it here. Make sure it is in AnnData format or pandas DataFrame (cells × genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Imputed data shape: (72650, 27837)\n",
      "             orig.ident  nCount_RNA  nFeature_RNA                 sample  \\\n",
      "43_01_73__s1  snParseBS     10080.0          3737  KPB25L-UV_Combination   \n",
      "43_01_92__s1  snParseBS      2913.0          1743  KPB25L-UV_Combination   \n",
      "43_01_94__s1  snParseBS     15986.0          5283  KPB25L-UV_Combination   \n",
      "43_02_41__s1  snParseBS      5901.0          2216  KPB25L-UV_Combination   \n",
      "43_02_56__s1  snParseBS     20456.0          5811  KPB25L-UV_Combination   \n",
      "\n",
      "             species  gene_count  tscp_count  mread_count  bc1_wind  bc2_wind  \\\n",
      "43_01_73__s1  GRCm39        3737       10080        15158        43         1   \n",
      "43_01_92__s1  GRCm39        1743        2913         4327        43         1   \n",
      "43_01_94__s1  GRCm39        5283       15986        23555        43         1   \n",
      "43_02_41__s1  GRCm39        2216        5901         8759        43         2   \n",
      "43_02_56__s1  GRCm39        5811       20456        30394        43         2   \n",
      "\n",
      "              ...  percent.mt percent.rb percent.hb integrated_snn_res.0.095  \\\n",
      "43_01_73__s1  ...    0.079365   0.327381   0.009921                        0   \n",
      "43_01_92__s1  ...    0.034329   0.137315   0.000000                        1   \n",
      "43_01_94__s1  ...    0.150131   0.337796   0.006255                        1   \n",
      "43_02_41__s1  ...    0.016946   0.169463   0.067785                        4   \n",
      "43_02_56__s1  ...    0.048885   0.386195   0.019554                        0   \n",
      "\n",
      "             seurat_clusters      sctype_custom    sctype_functional  \\\n",
      "43_01_73__s1               0            Unknown           Basal-like   \n",
      "43_01_92__s1               0       Cancer cells  Proliferating cells   \n",
      "43_01_94__s1               0       Cancer cells  Proliferating cells   \n",
      "43_02_41__s1               3  Endothelial cells    Endothelial cells   \n",
      "43_02_56__s1               0            Unknown           Basal-like   \n",
      "\n",
      "                 sctype_SCsubtype   final_annotation  \\\n",
      "43_01_73__s1           Basal-like       Cancer cells   \n",
      "43_01_92__s1  Proliferating cells       Cancer cells   \n",
      "43_01_94__s1  Proliferating cells       Cancer cells   \n",
      "43_02_41__s1    Endothelial cells  Endothelial cells   \n",
      "43_02_56__s1           Basal-like       Cancer cells   \n",
      "\n",
      "             final_annotation_functional  \n",
      "43_01_73__s1                  Basal-like  \n",
      "43_01_92__s1         Proliferating cells  \n",
      "43_01_94__s1         Proliferating cells  \n",
      "43_02_41__s1           Endothelial cells  \n",
      "43_02_56__s1                  Basal-like  \n",
      "\n",
      "[5 rows x 26 columns]\n"
     ]
    }
   ],
   "source": [
    "# Load imputed data from pickle file\n",
    "import pandas as pd\n",
    "import anndata as ad\n",
    "import pickle\n",
    "with open('./SimiCExampleRun/magic_output/magic_imputed_old.pickle', 'rb') as f:\n",
    "    imputed_data = pickle.load(f)\n",
    "print(f\"Imputed data shape: {imputed_data.shape}\")\n",
    "print(imputed_data.obs.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 2.2: Initialize Experiment Setup\n",
    "\n",
    "Create an experiment setup instance and directories:\n",
    "- `input_data`: Your imputed AnnData object or pandas DataFrame (cells × genes)\n",
    "- `tf_path`: Path to transcription factor list file (.csv or .txt)\n",
    "- `project_dir`: Directory where experiment files will be saved\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "\n",
    "<em><b>Note:</b> In case you do not a TF list, we provide a mouse TF list in the data folder that can be saved in your own data directory. \n",
    "TF mouse list was downloaded in December 2024 from \"https://guolab.wchscu.cn/AnimalTFDB4_static/download/TF_list_final/Mus_musculus_TF\"\n",
    "</em>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from importlib.resources import files\n",
    "p2tf = files(\"simicpipeline.data\").joinpath(\"Mus_musculus_TF.txt\")\n",
    "mouse_TF_df = pd.read_csv(p2tf, sep='\\t')\n",
    "mouse_TF = mouse_TF_df['Symbol']\n",
    "mouse_TF.to_csv('./data/TF_list.csv', index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Matrix shape: (72650, 27837)\n",
      "Number of cells: 72650\n",
      "Number of genes: 27837\n",
      "Number of TFs: 1611\n",
      "... Example TF names: ['Lin28b', 'Tbx2', 'Dmtf1', 'Irx4', 'Irf3']\n"
     ]
    }
   ],
   "source": [
    "# Initialize ExperimentSetup\n",
    "from simicpipeline import ExperimentSetup\n",
    "experiment = ExperimentSetup(\n",
    "    input_data = imputed_data, \n",
    "    tf_path = \"./data/TF_list.csv\", # Should have no header\n",
    "    project_dir='./SimiCExampleRun'\n",
    ")\n",
    "\n",
    "print(f\"Matrix shape: {experiment.matrix.shape}\")\n",
    "print(f\"Number of cells: {len(experiment.cell_names)}\")\n",
    "print(f\"Number of genes: {len(experiment.gene_names)}\")\n",
    "print(f\"Number of TFs: {len(experiment.tf_list)}\")\n",
    "print(f\"... Example TF names: {experiment.tf_list[0:5]}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The previous code automatically creates the SimiC directory structure:\n",
    "```\n",
    "project_dir/\n",
    "├── inputFiles/       # Input files for SimiC\n",
    "└── outputSimic/\n",
    "    ├── figures/      # For future visualizations\n",
    "    └── matrices/     # For future results\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 2.3: Calculate MAD and Select Genes\n",
    "\n",
    "Select top variable genes based on Median Absolute Deviation (MAD):\n",
    "- `n_tfs`: Number of top TF genes to select (default: 100)\n",
    "- `n_targets`: Number of top target genes to select (default: 1000)\n",
    "\n",
    "Returns a tuple of (TF_list, TARGET_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Removing 0 targets with MAD = 0\n",
      "Selecting top 1000 targets based on MAD.\n",
      "Selected 100 TFs\n",
      "Selected 1000 targets\n",
      "\n",
      "Top 10 TFs: ['Hmga2', 'Bnc2', 'Zfpm2', 'Satb2', 'Mecom', 'Zeb1', 'Glis3', 'Zeb2', 'Ebf1', 'Grhl2']\n",
      "\n",
      "Top 10 targets: ['Xist', 'Brinp3', 'Tenm4', '4930467D21Rik', 'Igf2bp2', 'Rad51b', 'Nop58', 'Gm49890', 'Pip5k1b', 'Ccbe1']\n"
     ]
    }
   ],
   "source": [
    "tf_list, target_list = experiment.calculate_mad_genes(\n",
    "    n_tfs=100,\n",
    "    n_targets=1000\n",
    ")\n",
    "\n",
    "print(f\"Selected {len(tf_list)} TFs\")\n",
    "print(f\"Selected {len(target_list)} targets\")\n",
    "print(f\"\\nTop 10 TFs: {tf_list[:10]}\")\n",
    "print(f\"\\nTop 10 targets: {target_list[:10]}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 2.4: Subset Data to Selected Genes\n",
    "\n",
    "Create a subset of your data containing only the selected TFs and targets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Subset data shape: (72650, 1100)\n"
     ]
    }
   ],
   "source": [
    "# Combine TF and target lists\n",
    "selected_genes = tf_list + target_list\n",
    "\n",
    "# Subset the data\n",
    "if isinstance(imputed_data, ad.AnnData):\n",
    "    subset_data = imputed_data[:, selected_genes].copy()\n",
    "elif isinstance(imputed_data, pd.DataFrame):\n",
    "    subset_data = imputed_data[selected_genes].copy()\n",
    "\n",
    "print(f\"Subset data shape: {subset_data.shape}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 2.5: Save Experiment Files\n",
    "\n",
    "Save the expression matrix and TF names in `.pickle` format and annotation file (optional) as `.txt`\n",
    "- `run_data`: `ad.AnnData` or `pd.Dataframe` with data to run in SimiC (Inputed and sliced according to experiment run)\n",
    "- `matrix_filename`: Filename to save `run_data` (saved with row/column headers)\n",
    "- `tf_filename`: Filename for TF names list for the experiment run.\n",
    "\n",
    "- `annotation`:`str` (Optional) if `run_data` is `ad.AnnData` and `annotation` is in `run_data.obs.columns`, it will create a `.txt` file with the phenotype annotations needed for SimiC with `index = False`, `header = False`\n",
    "\n",
    "All files are saved in the `inputFiles/` directory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Warning: Output file SimiCExampleRun/inputFiles/expression_matrix.pickle already exists and will be overwritten.\n",
      "Saved expression matrix to SimiCExampleRun/inputFiles/expression_matrix.pickle\n",
      "\n",
      "Warning: Output file SimiCExampleRun/inputFiles/TF_list.pickle already exists and will be overwritten.\n",
      "Saved 100 TFs to SimiCExampleRun/inputFiles/TF_list.pickle\n",
      "\n",
      "Warning: Annotation 'groups' not found in obs columns.\n",
      "\n",
      "Available columns:\n",
      " ['orig.ident', 'nCount_RNA', 'nFeature_RNA', 'sample', 'species', 'gene_count', 'tscp_count', 'mread_count', 'bc1_wind', 'bc2_wind', 'bc3_wind', 'bc1_well', 'bc2_well', 'bc3_well', 'treatment', 'cellLine', 'percent.mt', 'percent.rb', 'percent.hb', 'integrated_snn_res.0.095', 'seurat_clusters', 'sctype_custom', 'sctype_functional', 'sctype_SCsubtype', 'final_annotation', 'final_annotation_functional']\n",
      "Please manually provide an appropriate annotation file to SimiCPipeline in SimiCExampleRun/inputFiles\n",
      "\n",
      "-------\n",
      "\n",
      "Experiment files saved successfully.\n",
      "\n",
      "-------\n",
      "\n"
     ]
    }
   ],
   "source": [
    "experiment.save_experiment_files(\n",
    "    run_data = subset_data,\n",
    "    matrix_filename = 'expression_matrix.pickle',\n",
    "    tf_filename = 'TF_list.pickle',\n",
    "    annotation = 'groups'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Warning: Output file SimiCExampleRun/inputFiles/expression_matrix.pickle already exists and will be overwritten.\n",
      "Saved expression matrix to SimiCExampleRun/inputFiles/expression_matrix.pickle\n",
      "\n",
      "Warning: Output file SimiCExampleRun/inputFiles/TF_list.pickle already exists and will be overwritten.\n",
      "Saved 100 TFs to SimiCExampleRun/inputFiles/TF_list.pickle\n",
      "\n",
      "-------\n",
      "\n",
      "Annotation 'treatment' found in obs columns!\n",
      "Warning: annotation is not numeric. Will convert from categorical to numeric.\n",
      "\n",
      "Annotation distribution: {0: 17259, 1: 20491, 2: 15426, 3: 19474}\n",
      "\n",
      "Warning: Output file SimiCExampleRun/inputFiles/treatment_annotation.txt already exists and will be overwritten.\n",
      "Saved annotation to SimiCExampleRun/inputFiles/treatment_annotation.txt\n",
      "\n",
      "-------\n",
      "\n",
      "Experiment files saved successfully.\n",
      "\n",
      "-------\n",
      "\n"
     ]
    }
   ],
   "source": [
    "experiment.save_experiment_files(\n",
    "    run_data = subset_data,\n",
    "    matrix_filename = 'expression_matrix.pickle',\n",
    "    tf_filename = 'TF_list.pickle',\n",
    "    annotation = 'treatment'\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "<b>Success!</b> All preprocessing steps completed. Your files are ready for SimiC analysis.\n",
    "</div>\n",
    "\n",
    "## Step 2.7: Verify Saved Files\n",
    "\n",
    "Check that all files were created correctly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Input files directory: SimiCExampleRun/inputFiles\n",
      "\n",
      "Files created:\n",
      "  - TF_list.pickle (0.79 KB)\n",
      "  - treatment_annotation.txt (141.89 KB)\n",
      "  - Tumor_KPB25L_expression_matrix.pickle (625410.01 KB)\n",
      "  - expression_matrix.pickle (625410.01 KB)\n",
      "\n",
      "-------\n",
      "\n",
      "AnnData object with n_obs × n_vars = 72650 × 1100\n",
      "    obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'sample', 'species', 'gene_count', 'tscp_count', 'mread_count', 'bc1_wind', 'bc2_wind', 'bc3_wind', 'bc1_well', 'bc2_well', 'bc3_well', 'treatment', 'cellLine', 'percent.mt', 'percent.rb', 'percent.hb', 'integrated_snn_res.0.095', 'seurat_clusters', 'sctype_custom', 'sctype_functional', 'sctype_SCsubtype', 'final_annotation', 'final_annotation_functional'\n",
      "    var: 'vf_vst_counts.1_mean', 'vf_vst_counts.1_variance', 'vf_vst_counts.1_variance.expected', 'vf_vst_counts.1_variance.standardized', 'vf_vst_counts.1_variable', 'vf_vst_counts.1_rank', 'vf_vst_counts.2_mean', 'vf_vst_counts.2_variance', 'vf_vst_counts.2_variance.expected', 'vf_vst_counts.2_variance.standardized', 'vf_vst_counts.2_variable', 'vf_vst_counts.2_rank', 'vf_vst_counts.3_mean', 'vf_vst_counts.3_variance', 'vf_vst_counts.3_variance.expected', 'vf_vst_counts.3_variance.standardized', 'vf_vst_counts.3_variable', 'vf_vst_counts.3_rank', 'vf_vst_counts.4_mean', 'vf_vst_counts.4_variance', 'vf_vst_counts.4_variance.expected', 'vf_vst_counts.4_variance.standardized', 'vf_vst_counts.4_variable', 'vf_vst_counts.4_rank', 'vf_vst_counts.5_mean', 'vf_vst_counts.5_variance', 'vf_vst_counts.5_variance.expected', 'vf_vst_counts.5_variance.standardized', 'vf_vst_counts.5_variable', 'vf_vst_counts.5_rank', 'vf_vst_counts.6_mean', 'vf_vst_counts.6_variance', 'vf_vst_counts.6_variance.expected', 'vf_vst_counts.6_variance.standardized', 'vf_vst_counts.6_variable', 'vf_vst_counts.6_rank', 'vf_vst_counts.7_mean', 'vf_vst_counts.7_variance', 'vf_vst_counts.7_variance.expected', 'vf_vst_counts.7_variance.standardized', 'vf_vst_counts.7_variable', 'vf_vst_counts.7_rank', 'vf_vst_counts.8_mean', 'vf_vst_counts.8_variance', 'vf_vst_counts.8_variance.expected', 'vf_vst_counts.8_variance.standardized', 'vf_vst_counts.8_variable', 'vf_vst_counts.8_rank', 'vf_vst_counts_mean', 'vf_vst_counts_variance', 'vf_vst_counts_variance.expected', 'vf_vst_counts_variance.standardized', 'vf_vst_counts_variable', 'vf_vst_counts_rank', 'var.features', 'var.features.rank'\n",
      "\n",
      "-------\n",
      "\n",
      "             orig.ident  nCount_RNA  nFeature_RNA                 sample  \\\n",
      "43_01_73__s1  snParseBS     10080.0          3737  KPB25L-UV_Combination   \n",
      "43_01_92__s1  snParseBS      2913.0          1743  KPB25L-UV_Combination   \n",
      "43_01_94__s1  snParseBS     15986.0          5283  KPB25L-UV_Combination   \n",
      "43_02_41__s1  snParseBS      5901.0          2216  KPB25L-UV_Combination   \n",
      "43_02_56__s1  snParseBS     20456.0          5811  KPB25L-UV_Combination   \n",
      "\n",
      "             species  gene_count  tscp_count  mread_count  bc1_wind  bc2_wind  \\\n",
      "43_01_73__s1  GRCm39        3737       10080        15158        43         1   \n",
      "43_01_92__s1  GRCm39        1743        2913         4327        43         1   \n",
      "43_01_94__s1  GRCm39        5283       15986        23555        43         1   \n",
      "43_02_41__s1  GRCm39        2216        5901         8759        43         2   \n",
      "43_02_56__s1  GRCm39        5811       20456        30394        43         2   \n",
      "\n",
      "              ...  percent.mt percent.rb percent.hb integrated_snn_res.0.095  \\\n",
      "43_01_73__s1  ...    0.079365   0.327381   0.009921                        0   \n",
      "43_01_92__s1  ...    0.034329   0.137315   0.000000                        1   \n",
      "43_01_94__s1  ...    0.150131   0.337796   0.006255                        1   \n",
      "43_02_41__s1  ...    0.016946   0.169463   0.067785                        4   \n",
      "43_02_56__s1  ...    0.048885   0.386195   0.019554                        0   \n",
      "\n",
      "             seurat_clusters      sctype_custom    sctype_functional  \\\n",
      "43_01_73__s1               0            Unknown           Basal-like   \n",
      "43_01_92__s1               0       Cancer cells  Proliferating cells   \n",
      "43_01_94__s1               0       Cancer cells  Proliferating cells   \n",
      "43_02_41__s1               3  Endothelial cells    Endothelial cells   \n",
      "43_02_56__s1               0            Unknown           Basal-like   \n",
      "\n",
      "                 sctype_SCsubtype   final_annotation  \\\n",
      "43_01_73__s1           Basal-like       Cancer cells   \n",
      "43_01_92__s1  Proliferating cells       Cancer cells   \n",
      "43_01_94__s1  Proliferating cells       Cancer cells   \n",
      "43_02_41__s1    Endothelial cells  Endothelial cells   \n",
      "43_02_56__s1           Basal-like       Cancer cells   \n",
      "\n",
      "             final_annotation_functional  \n",
      "43_01_73__s1                  Basal-like  \n",
      "43_01_92__s1         Proliferating cells  \n",
      "43_01_94__s1         Proliferating cells  \n",
      "43_02_41__s1           Endothelial cells  \n",
      "43_02_56__s1                  Basal-like  \n",
      "\n",
      "[5 rows x 26 columns]\n"
     ]
    }
   ],
   "source": [
    "input_dir = experiment.input_files_dir\n",
    "print(f\"Input files directory: {input_dir}\")\n",
    "print(f\"\\nFiles created:\")\n",
    "for file in os.listdir(input_dir):\n",
    "    filepath = input_dir / file\n",
    "    size = os.path.getsize(filepath) / 1024  # Size in KB\n",
    "    print(f\"  - {file} ({size:.2f} KB)\")\n",
    "\n",
    "# Preview the inputed expression matrix\n",
    "print(\"\\n-------\\n\")\n",
    "print(subset_data)\n",
    "print(\"\\n-------\\n\")\n",
    "print(subset_data.obs[0:5])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   0\n",
      "0  0\n",
      "1  0\n",
      "2  0\n",
      "3  0\n",
      "4  0\n",
      "\n",
      "(72650, 1)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0\n",
       "1    20491\n",
       "3    19474\n",
       "0    17259\n",
       "2    15426\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "annot = pd.read_csv(input_dir / 'treatment_annotation.txt',header=None)\n",
    "print(annot.head())\n",
    "\n",
    "print(f\"\\n{annot.shape}\")\n",
    "annot[0].value_counts()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Summary\n",
    "\n",
    "This tutorial covered:\n",
    "\n",
    "✓ Loading and filtering scRNA-seq data\n",
    "\n",
    "✓ Running MAGIC imputation\n",
    "\n",
    "✓ Selecting top variable genes using MAD\n",
    "\n",
    "✓ Preparing input files for SimiC analysis with proper directory structure\n",
    "\n",
    "### Output Directory Structure\n",
    "\n",
    "Your output directory now contains:\n",
    "```\n",
    "SimicExampleRun/\n",
    "├── magic_output/\n",
    "│   └── magic_imputed.pickle\n",
    "├── inputFiles/\n",
    "│   ├── expression_matrix.pickle\n",
    "│   ├── TF_list.pickle\n",
    "│   └── groups_phenotype.txt\n",
    "└── outputSimic/\n",
    "    ├── figures/\n",
    "    └── matrices/\n",
    "```\n",
    "## Alternative Approach\n",
    "\n",
    "In this tutorial we used the whole Magic-inputed matrix (obtained in #Part 1: MAGIC Imputation Pipeline) and selected top MAD genes but you may want to run SimiC in a subset of cells from your data. Make sure you slice the data before you inilitalize the `ExperimentSetup` class so MAD genes are calculated over your cells of interest.\n",
    "\n",
    "All these steps are optional but recommended before running SimiCPipeline just:\n",
    "1. Make sure to **prepare cell assignments** if not done before: Create a file with cell phenotype labels matching  the same order as your expression matrix that you will use in SimiC.\n",
    "2. Prepare **expression matrix** pickle file\n",
    "3. Prepare **TF_list** pickle with TF genes in your expression matrix.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_mask = imputed_data.obs['cellLine'].isin([\"KPB25L\"]) & imputed_data.obs['final_annotation'].isin([\"Cancer cells\"])\n",
    "sum(cell_mask)\n",
    "subset_inputed_data = imputed_data[cell_mask,:].copy()\n",
    "subset_inputed_data.shape\n",
    "\n",
    "experiment = ExperimentSetup(\n",
    "    input_data = subset_inputed_data, \n",
    "    tf_path = \"./data/TF_list.csv\", # Should have no header\n",
    "    project_dir='./SimiCExampleRun/KPB25L/Tumor'\n",
    ")\n",
    "# Then follow the same steps as above to select genes and save experiment files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The output directory will then look like \n",
    "```\n",
    "SimicExampleRun/\n",
    "├── magic_output/\n",
    "│   └── magic_imputed.pickle\n",
    "├── inputFiles/\n",
    "│   ├── expression_matrix.pickle\n",
    "│   ├── TF_list.pickle\n",
    "│   └── groups_phenotype.txt\n",
    "├── outputSimic/\n",
    "│   ├── figures/\n",
    "│   └── matrices/\n",
    "├── KPB25L/\n",
    "    └──Tumor/\n",
    "        ├── inputFiles/\n",
    "        │   ├── expression_matrix.pickle\n",
    "        │   ├── TF_list.pickle\n",
    "        │   └── groups_phenotype.txt\n",
    "        └── outputSimic/\n",
    "            ├── figures/\n",
    "            └── matrices/\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Next steps:\n",
    "1. **Run SimiC**: Use `SimiCPipeline` class to run SimiC.\n",
    "2. **Explore results**: Use `SimicVisualization`class to analyze GRNs and TF activities.\n",
    "\n",
    "Check `Tutorial_SimiCPipeline_simple.ipynb` or `Tutorial_SimiCPipeline_full` for more info.\n",
    "\n",
    "\n",
    "### Final Notes\n",
    "\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "<b>Data Format:</b> All matrices are stored as cells × genes (rows = cells, columns = genes)\n",
    "</div>\n",
    "\n",
    "<div class=\"alert alert-block alert-warning\">\n",
    "<b>Memory Usage:</b> MAGIC imputation can be memory-intensive for large datasets. Consider using a machine with sufficient RAM and adjusting MAGIC parameters (n_jobs, knn, t)\n",
    "</div>\n",
    "\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "<b>Please note:</b> Although you will be able to pass custom file/direcotry paths, we highly recommend to follow the directory structure described above and follow this tutorial before running SimiC to avoid errors.\n",
    "</div>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SimiC Pipeline Kernel",
   "language": "python",
   "name": "simicpipeline"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
