{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SimiC Preprocessing Pipeline Tutorial\n",
    "\n",
    ">*Author: Irene Marín-Goñi, PhD student - ML4BM group (CIMA University of Navarra)*\n",
    "\n",
    "This notebook demonstrates how to preprocess single-cell RNA-seq data for SimiC analysis.\n",
    "\n",
    "## Overview\n",
    "This preprocessing tutorial covers:\n",
    "1. Package installation and setup\n",
    "2. MAGIC imputation pipeline\n",
    "3. Gene selection and experiment setup\n",
    "4. Preparing input files for SimiC\n",
    "\n",
    "For running SimiC analysis see `Tutorial_SimiCPipeline_full.ipynb`\n",
    "\n",
    "## Introduction\n",
    "Before running SimiC, you need to:\n",
    "- Impute your scRNA-seq data. We recommend to use [MAGIC](https://pypi.org/project/magic-impute/3.0.0/) and include a wrapper class `MagicPipeline` to ease the process.\n",
    "- Select top variable genes based on Median Absolute Deviation (MAD) or the genes of interest from which you want to infer the gene regulatory network.\n",
    "- Prepare input files in the correct format for SimiCPipeline.\n",
    "\n",
    "This tutorial shows you how to do all of this using the SimiC preprocessing modules."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup\n",
    "\n",
    "The easiest way to configure your environment is to follow the `README` instructions using `poetry` (or `Docker`).\n",
    "\n",
    "\n",
    "Required packages for this tutorial:\n",
    "- simicpipeline\n",
    "- anndata\n",
    "- pandas\n",
    "- numpy\n",
    "- os\n",
    "- pickle\n",
    "\n",
    "Internally simicpipeline also uses:\n",
    "- scipy\n",
    "- sklearn\n",
    "- scprep (in preprocessing)\n",
    "- magic-impute (in preprocessing)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import Modules\n",
    "First, import the necessary preprocessing modules."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/workdir\n",
      "['data']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SimiC pipeline version:  {'0.1.0'}\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "print(os.getcwd())\n",
    "print(os.listdir())\n",
    "import simicpipeline \n",
    "print(f\"SimiC pipeline version: \", {simicpipeline.__version__})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='part1'></a>\n",
    "# Part 1: MAGIC Imputation Pipeline\n",
    "\n",
    "MAGIC (Markov Affinity-based Graph Imputation of Cells) is used to denoise and impute scRNA-seq data. `MagicPipeline` facilitates the steps described in [Magic Tutorial](\"https://magic.readthedocs.io/en/stable/tutorial.html\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Step 1.1: Load Your Data\n",
    "\n",
    "Load your raw expression data. Note that `MacigPipeline` class expects AnnData format.\n",
    "\n",
    "Below you will see different examples on how to generate the AnnData object from different input files (including Seurat if you are more familiar with R)\n",
    "\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "<em> <b>Note 1:</b> If you have already filtered and transformed your data but want to repeat following these steps, make sure the adata object has the raw counts in the `adata.raw.X` slot</b>\n",
    "</div>\n",
    "\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "<b>Note 2:</b> If you have already processed (inputed) your data  you can jump to Part 2 Experiment Setup\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### From Seurat\n",
    "If you have your data in a Seurat object (R package) the easiest approach is this:\n",
    "```r\n",
    "\n",
    "write.table(data.frame(Cells = colnames(seurat_obj)), file.path(\"/path/to/data/\", \"cell_ids.txt\"), row.names = FALSE, col.names = FALSE, quote = FALSE)\n",
    "\n",
    "write.table(data.frame(Genes = rownames(seurat_obj)), file.path(\"/path/to/data/\", \"genes_ids.txt\"), row.names = FALSE, col.names = FALSE, quote = FALSE)\n",
    "\n",
    "write.table(seurat_obj@metadata, file.path(\"/path/to/data/\", \"metadata.csv\"), sep = \",\",row.names = TRUE, col.names = TRUE, quote = FALSE)\n",
    "\n",
    "m_raw = GetAssayData(seurat_obj, assay = \"RNA\", layer = \"counts\") # Take the raw counts\n",
    "\n",
    "Matrix::writeMM(m_raw, paste0(magic_path, \"/singlecell_matrix.mtx\")) # Save it in MatrixMarket format \n",
    "```\n",
    "\n",
    "Then follow the example code below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Load your AnnData object here\n",
      "(72650, 8)\n",
      "                             sample    treatment  cell_line  \\\n",
      "cell                                                          \n",
      "43_01_92__s1  KPB25L-UV_Combination  Combination  KPB25L-UV   \n",
      "43_01_73__s1  KPB25L-UV_Combination  Combination  KPB25L-UV   \n",
      "43_01_94__s1  KPB25L-UV_Combination  Combination  KPB25L-UV   \n",
      "43_02_41__s1  KPB25L-UV_Combination  Combination  KPB25L-UV   \n",
      "43_02_56__s1  KPB25L-UV_Combination  Combination  KPB25L-UV   \n",
      "\n",
      "               final_annotation final_annotation_functional  \\\n",
      "cell                                                          \n",
      "43_01_92__s1       Cancer cells         Proliferating cells   \n",
      "43_01_73__s1       Cancer cells                  Basal-like   \n",
      "43_01_94__s1       Cancer cells         Proliferating cells   \n",
      "43_02_41__s1  Endothelial cells           Endothelial cells   \n",
      "43_02_56__s1       Cancer cells                  Basal-like   \n",
      "\n",
      "                nn_majority_label  nn_majority_frac  flag_misplaced  \n",
      "cell                                                                 \n",
      "43_01_92__s1  Proliferating cells               1.0           False  \n",
      "43_01_73__s1           Basal-like               1.0           False  \n",
      "43_01_94__s1  Proliferating cells               1.0           False  \n",
      "43_02_41__s1    Endothelial cells               1.0           False  \n",
      "43_02_56__s1           Basal-like               0.9           False  \n"
     ]
    }
   ],
   "source": [
    "\n",
    "print(\"Load your AnnData object here\")\n",
    "\n",
    "# # Example: Load from Matrix Market format\n",
    "import pandas as pd\n",
    "import anndata as ad\n",
    "from pathlib import Path\n",
    "\n",
    "# This will return a pd.DataFrame\n",
    "df = simicpipeline.load_from_matrix_market( \n",
    "    matrix_path=Path(\"./data/singlecell_matrix.mtx\"),\n",
    "    genes_path=Path(\"./data/genes_ids.txt\"),\n",
    "    cells_path=Path(\"./data/cell_ids.txt\"),\n",
    "    transpose=True,\n",
    "    cells_index_name=\"Cell\",\n",
    ")\n",
    "adata = ad.AnnData(X=df.values, obs=pd.DataFrame(index=df.index), var=pd.DataFrame(index=df.columns))\n",
    "\n",
    "obs_meta=pd.read_csv(\"/path/to/data/metadata.csv\", sep = \"\\.\", index_col=0)\n",
    "print(obs_meta.shape)\n",
    "print(obs_meta.head())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "<b>WARNING!!:</b> Make sure that the index of obs_meta matches the cell names in df.index. Otherwise it will lead to misalignment of metadata and expression data.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Match the obs metadata index to the adata.obs_names\n",
    "obs_meta = obs_meta.loc[adata.obs_names]\n",
    "adata = ad.AnnData(adata.X, obs=obs_meta, var=adata.var)\n",
    "# If your data is raw, you should set it properly in AnnData object\n",
    "adata.raw = adata.copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alternative loading options"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(\"Load your AnnData object here\")\n",
    "\n",
    "# # Example: Load from CSV files\n",
    "# import pandas as pd\n",
    "# import anndata as ad\n",
    "# # This example assumes that in csv rows are cells and columns are genes\n",
    "# expression_data = pd.read_csv('path/to/expression_data.csv', index_col = 0) # Column 1 as cell IDs\n",
    "# # Match the observation metadata\n",
    "# metadata = pd.read_csv('path/to/metadata.csv', index_col=0)\n",
    "# metadata = metadata.loc[expression_data.index]\n",
    "# adata = ad.AnnData(X=expression_data.values, obs=metadata)\n",
    "\n",
    "# # Example: Load from 10X format\n",
    "# adata = ad.read_10x_mtx('path/to/10x/directory')\n",
    "\n",
    "# # Example: Load from h5ad file using simicpipeline function\n",
    "# adata = simicpipeline.load_from_anndata('path/to/your/data.h5ad')\n",
    "\n",
    "# If your data is raw, you should set it properly with\n",
    "# print(hasattr(adata, 'raw'))\n",
    "# adata.raw = adata.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 72650 × 36774\n",
       "    obs: 'sample', 'treatment', 'cell_line', 'final_annotation', 'final_annotation_functional', 'nn_majority_label', 'nn_majority_frac', 'flag_misplaced'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1.2: Initialize MAGIC Pipeline\n",
    "\n",
    "Create a MAGIC pipeline instance:\n",
    "- `input_data`: Your AnnData object. If you run the full pipline starting from raw counts they should be in `adata.raw.X`\n",
    "- `project_dir`: Project directory where `magic_output`dir will be created and files will be saved\n",
    "- `magic_output_file`: Filename for the imputed data (default: 'magic_data_allcells_sqrt.pickle')\n",
    "- `filtered`: Set to True if data is already filtered (low quality cells and genes) (default: False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating project directory: SimiCExampleRun\n",
      "MagicPipeline(\n",
      "  data = AnnData object with (n_obs × n_vars) = 72650 × 36774,\n",
      "  filtered = False,\n",
      "  imputed = False,\n",
      "  magic_data = None,\n",
      "  project_dir = 'SimiCExampleRun'\n",
      ")\n"
     ]
    }
   ],
   "source": [
    "# This command will initialize the MAGIC pipeline and generate the output directory if it does not exist\n",
    "from simicpipeline import MagicPipeline\n",
    "magic_pipeline = MagicPipeline(\n",
    "    input_data= adata,\n",
    "    project_dir='./SimiCExampleRun',\n",
    "    magic_output_file='magic_imputed.pickle',\n",
    "    filtered=False\n",
    ")\n",
    "\n",
    "print(magic_pipeline)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SimiCExampleRun/\n",
      "└── magic_output/\n"
     ]
    }
   ],
   "source": [
    "magic_pipeline.print_project_info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1.3: Filter Cells and Genes\n",
    "\n",
    "Remove low-quality cells and lowly-expressed genes:\n",
    "- `min_cells_per_gene`: Minimum number of cells expressing a gene (default: 10)\n",
    "- `min_umis_per_cell`: Minimum total UMI counts per cell (default: 500)\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "<em><b>Note:</b> If your data was already filtered you can skip this step and set the flitered argument flag to `True` in the previous step.</em>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Filtering cells and genes...\n",
      "Before filtering: 72650 cells x 36774 genes\n",
      "Keeping 27837/36774 genes (75.70%)\n",
      "Keeping 72650/72650 cells (100.00%)\n",
      "All cells pass the filter!\n",
      "After filtering: 72650 cells x 27837 genes\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "MagicPipeline(\n",
       "  data = AnnData object with (n_obs × n_vars) = 72650 × 27837,\n",
       "  filtered = True,\n",
       "  imputed = False,\n",
       "  magic_data = None,\n",
       "  project_dir = 'SimiCExampleRun'\n",
       ")"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "magic_pipeline.filter_cells_and_genes(min_cells_per_gene = 10, min_umis_per_cell = 500)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1.4: Normalize Data\n",
    "\n",
    "Perform library size normalization with `scprep` followed by square root transformation.\n",
    "\n",
    "<div class=\"alert alert-block alert-warning\">\n",
    "<em><b>Note:</b> this will overide adata.X with normalized data and remove adata.raw slot.</em>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Normalizing data...\n",
      "After normalization: 72650 cells x 27837 genes\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "MagicPipeline(\n",
       "  data = AnnData object with (n_obs × n_vars) = 72650 × 27837,\n",
       "  filtered = True,\n",
       "  imputed = False,\n",
       "  magic_data = None,\n",
       "  project_dir = 'SimiCExampleRun'\n",
       ")"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Note this will overide adata.X with normalized data and remove adata.raw slot\n",
    "magic_pipeline.normalize_data()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1.5: Run MAGIC Imputation\n",
    "Run magic imputation with defaul parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Running MAGIC imputation...\n",
      "Calculating MAGIC...\n",
      "  Running MAGIC on 72650 cells and 27837 genes.\n",
      "  Calculating graph and diffusion operator...\n",
      "    Calculating PCA...\n",
      "    Calculated PCA in 83.78 seconds.\n",
      "    Calculating KNN search...\n",
      "    Calculated KNN search in 2.98 seconds.\n",
      "    Calculating affinities...\n",
      "    Calculated affinities in 7.15 seconds.\n",
      "  Calculated graph and diffusion operator in 94.08 seconds.\n",
      "  Running MAGIC with `solver='exact'` on 27837-dimensional data may take a long time. Consider denoising specific genes with `genes=<list-like>` or using `solver='approximate'`.\n",
      "  Calculating imputation...\n",
      "  Calculated imputation in 187.77 seconds.\n",
      "Calculated MAGIC in 283.46 seconds.\n",
      "MAGIC imputation complete:  72650 cells x 27837 genes\n",
      "\n",
      "Saving MAGIC-imputed data to SimiCExampleRun/magic_output/magic_imputed.pickle\n",
      "Saved successfully to SimiCExampleRun/magic_output/magic_imputed.pickle\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "MagicPipeline(\n",
       "  data = AnnData object with (n_obs × n_vars) = 72650 × 27837,\n",
       "  filtered = True,\n",
       "  imputed = True,\n",
       "  magic_data = AnnData object with n_obs × n_vars = 72650 × 27837,\n",
       "  project_dir = 'SimiCExampleRun'\n",
       ")"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "magic_pipeline.run_magic(\n",
    "    random_state=123,\n",
    "    n_jobs=-2,  # Use all but 1 CPU cores\n",
    "    save_data=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [],
   "source": [
    "magic_pipeline.magic_adata.write_h5ad(magic_pipeline.magic_output_file.with_suffix('.h5ad'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "If you want to run MAGIC imputation with custom parameters you can pass them as `**kwargs`:\n",
    "- `t`: Number of diffusion steps (default: 'auto')\n",
    "- `knn`: Number of nearest neighbors (default: 5)\n",
    "- `decay`: Decay rate for kernel (default: 1)\n",
    "- `n_jobs`: Number of parallel jobs (default: -2)\n",
    "- `genes`: Genes to be returned. If None or \"all genes\" it returns teh entire matrix.\n",
    "- `save_data`: Whether to automatically save imputed data (default: True). If magic_output_file extension is .pickle will save it in .pickle, if h5ad, will save in adata format.\n",
    "\n",
    "See [MAGIC documentation](https://magic.readthedocs.io/) for more parameter options."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1.6: Check Pipeline Status"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SimiCExampleRun/\n",
      "└── magic_output/\n",
      "    ├── magic_imputed.h5ad\n",
      "    └── magic_imputed.pickle\n"
     ]
    }
   ],
   "source": [
    "magic_pipeline.print_project_info(max_depth=2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "<b>Success!</b> MAGIC imputation is complete. The imputed data is saved in the magic_output directory.\n",
    "</div>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part 2: Experiment Setup and Gene Selection\n",
    "<a id='experiment'></a>\n",
    "\n",
    "Now we will select top variable genes and prepare input files for SimiCPipeline. \n",
    "<div class=\"alert alert-block alert-info\">\n",
    "\n",
    "<em><b>Note:</b> if you have your data filtered, normalized and imputed with alternative methods you can start from here.</em>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Step 2.1: Load Imputed Data\n",
    "\n",
    "In this example we will start from the imputed AnnData object from the MAGIC pipeline. If you saved and stopped your work, you can re-load the object with the following code:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# If you saved it in h5ad format, you can load it back using:\n",
    "# import simicpipeline\n",
    "# imputed_data = simicpipeline.load_from_anndata('./SimiCExampleRun/magic_output/magic_imputed.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # If you saved it in pickle format, you can load it back using:\n",
    "# import pickle\n",
    "# with open('./SimiCExampleRun/magic_output/magic_imputed.pickle', 'rb') as f:\n",
    "#     imputed_data = pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 72650 × 27837\n",
       "    obs: 'sample', 'treatment', 'cell_line', 'final_annotation', 'final_annotation_functional', 'nn_majority_label', 'nn_majority_frac', 'flag_misplaced'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# If you contintue from the previous section you can access the MAGIC-imputed AnnData object\n",
    "imputed_data = magic_pipeline.magic_adata.copy()\n",
    "imputed_data\n",
    "# If you have inputed data from other methods, you can load it here. Make sure it is in AnnData format or pandas DataFrame (cells × genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Imputed data shape: (72650, 27837)\n",
      "                             sample    treatment  cell_line  \\\n",
      "Cell                                                          \n",
      "43_01_73__s1  KPB25L-UV_Combination  Combination  KPB25L-UV   \n",
      "43_01_92__s1  KPB25L-UV_Combination  Combination  KPB25L-UV   \n",
      "43_01_94__s1  KPB25L-UV_Combination  Combination  KPB25L-UV   \n",
      "43_02_41__s1  KPB25L-UV_Combination  Combination  KPB25L-UV   \n",
      "43_02_56__s1  KPB25L-UV_Combination  Combination  KPB25L-UV   \n",
      "\n",
      "               final_annotation final_annotation_functional  \\\n",
      "Cell                                                          \n",
      "43_01_73__s1       Cancer cells                  Basal-like   \n",
      "43_01_92__s1       Cancer cells         Proliferating cells   \n",
      "43_01_94__s1       Cancer cells         Proliferating cells   \n",
      "43_02_41__s1  Endothelial cells           Endothelial cells   \n",
      "43_02_56__s1       Cancer cells                  Basal-like   \n",
      "\n",
      "                nn_majority_label  nn_majority_frac  flag_misplaced  \n",
      "Cell                                                                 \n",
      "43_01_73__s1           Basal-like               1.0           False  \n",
      "43_01_92__s1  Proliferating cells               1.0           False  \n",
      "43_01_94__s1  Proliferating cells               1.0           False  \n",
      "43_02_41__s1    Endothelial cells               1.0           False  \n",
      "43_02_56__s1           Basal-like               0.9           False  \n"
     ]
    }
   ],
   "source": [
    "print(f\"Imputed data shape: {imputed_data.shape}\")\n",
    "print(imputed_data.obs.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 2.2: Initialize Experiment Setup\n",
    "\n",
    "Create an experiment setup instance and directories:\n",
    "- `input_data`: Your imputed AnnData object or pandas DataFrame (cells × genes)\n",
    "- `tf_path`: Path to transcription factor (TF) list file (.csv or .txt)\n",
    "- `project_dir`: Directory where experiment files will be saved\n",
    "\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "\n",
    "<em><b>Note:</b> In case you do not have a TF list: \n",
    "</em>\n",
    "</div>\n",
    "\n",
    "We provide a mouse TF list in the data folder that can be saved in your working data directory. \n",
    "- TF mouse list was downloaded in December 2024 from [AnimalTFDB4](https://guolab.wchscu.cn/AnimalTFDB4_static/download/TF_list_final/Mus_musculus_TF)\n",
    "- TF human list was downloaded in February 2026 from [XX](#)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this tutorial we are working wiht mouse data so we will use the TF list from `AnimalTFDB4`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from importlib.resources import files\n",
    "import pandas as pd\n",
    "p2tf = files(\"simicpipeline.data\").joinpath(\"Mus_musculus_TF.txt\")\n",
    "mouse_TF_df = pd.read_csv(p2tf, sep='\\t')\n",
    "mouse_TF = mouse_TF_df['Symbol']\n",
    "mouse_TF.to_csv('./data/TF_list.csv', index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Matrix shape: (72650, 27837)\n",
      "Number of cells: 72650\n",
      "Number of genes: 27837\n",
      "Number of TFs: 1611\n",
      "... Example TF names: ['Lin28b', 'Tbx2', 'Dmtf1', 'Irx4', 'Irf3']\n",
      "\n",
      "\n",
      "======================================================================\n",
      "Current directory status\n",
      "======================================================================\n",
      "\n",
      "SimiCExampleRun/\n",
      "├── inputFiles/\n",
      "├── magic_output/\n",
      "|   ├── magic_imputed.h5ad\n",
      "|   └── magic_imputed.pickle\n",
      "└── outputSimic/\n",
      "    ├── figures/\n",
      "    └── matrices/\n"
     ]
    }
   ],
   "source": [
    "# Initialize ExperimentSetup\n",
    "from simicpipeline import ExperimentSetup\n",
    "experiment = ExperimentSetup(\n",
    "    input_data = imputed_data, \n",
    "    tf_path = \"./data/TF_list.csv\", # Should have no header\n",
    "    project_dir='./SimiCExampleRun'\n",
    ")\n",
    "\n",
    "print(f\"Matrix shape: {experiment.matrix.shape}\")\n",
    "print(f\"Number of cells: {len(experiment.cell_names)}\")\n",
    "print(f\"Number of genes: {len(experiment.gene_names)}\")\n",
    "print(f\"Number of TFs: {len(experiment.tf_list)}\")\n",
    "print(f\"... Example TF names: {experiment.tf_list[0:5]}\\n\")\n",
    "print(\"\\n\" + \"=\"*70)\n",
    "print(f\"Current directory status\")\n",
    "print(\"=\"*70 + \"\\n\")\n",
    "experiment.print_project_info(max_depth=2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The previous code automatically creates the SimiC directory structure:\n",
    "```\n",
    "project_dir/\n",
    "├── inputFiles/       # Input files for SimiC\n",
    "└── outputSimic/      # Output files generated by SimiCPipeline\n",
    "    ├── figures/      # For future visualizations\n",
    "    └── matrices/     # For future results\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 2.3: Calculate MAD and Select Genes\n",
    "\n",
    "Select top variable genes based on Median Absolute Deviation (MAD):\n",
    "- `n_tfs`: Number of top TF genes to select (default: 100)\n",
    "- `n_targets`: Number of top target genes to select (default: 1000)\n",
    "\n",
    "Returns a tuple of (TF_list, TARGET_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Removing 0 targets with MAD = 0\n",
      "Selecting top 1000 targets based on MAD.\n",
      "Selected 100 TFs\n",
      "Selected 1000 targets\n",
      "\n",
      "Top 10 TFs: ['Hmga2', 'Zfpm2', 'Bnc2', 'Glis3', 'Zeb2', 'Pbx1', 'Ebf1', 'Zeb1', 'Mecom', 'Nfib']\n",
      "\n",
      "Top 10 targets: ['Rn18s-rs5', 'Malat1', 'Cmss1', 'Lamp2', 'Brinp3', 'Rad51b', 'Ccbe1', 'Tenm4', 'Nop58', 'Xist']\n"
     ]
    }
   ],
   "source": [
    "tf_list, target_list = experiment.calculate_mad_genes(\n",
    "    n_tfs=100,\n",
    "    n_targets=1000\n",
    ")\n",
    "\n",
    "print(f\"Selected {len(tf_list)} TFs\")\n",
    "print(f\"Selected {len(target_list)} targets\")\n",
    "print(f\"\\nTop 10 TFs: {tf_list[:10]}\")\n",
    "print(f\"\\nTop 10 targets: {target_list[:10]}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 2.4: Subset Data to Selected Genes\n",
    "\n",
    "Create a subset of your data containing only the selected TFs and targets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Subset data shape: (72650, 1100)\n"
     ]
    }
   ],
   "source": [
    "# Combine TF and target lists\n",
    "import anndata as ad\n",
    "selected_genes = tf_list + target_list\n",
    "\n",
    "# Subset the data\n",
    "if isinstance(imputed_data, ad.AnnData):\n",
    "    subset_data = imputed_data[:, selected_genes].copy()\n",
    "elif isinstance(imputed_data, pd.DataFrame):\n",
    "    subset_data = imputed_data[selected_genes].copy()\n",
    "\n",
    "print(f\"Subset data shape: {subset_data.shape}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 2.5: Save Experiment Files\n",
    "\n",
    "Save the expression matrix and TF names in `.pickle` format and annotation file (optional) as `.txt`\n",
    "- `run_data`: `ad.AnnData` or `pd.Dataframe` with data to run in SimiC (Inputed and sliced according to experiment run)\n",
    "- `matrix_filename`: Filename to save `run_data` (saved with row/column headers). Can be `.pickle`or `csv`.\n",
    "- `tf_filename`: Filename for TF names list for the experiment run. Can be `.pickle`or `csv`. Even though you have a general TF_list file, this function will save the TFs selected by MAD that are found in your run_data.\n",
    "\n",
    "- `annotation`:`str` (Optional) if `run_data` is `ad.AnnData` and `annotation` is in `run_data.obs.columns`, it will create a `.txt` file with the phenotype annotations needed for SimiC with `index = False`, `header = False`\n",
    "\n",
    "All files are saved in the `inputFiles/` directory.\n",
    "\n",
    "We recommend saving it in pickle format for fast load/dump process and save disk space."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved expression matrix to SimiCExampleRun/inputFiles/expression_matrix.csv\n",
      "Saved 100 TFs to SimiCExampleRun/inputFiles/TF_list.csv\n",
      "\n",
      "Warning: Annotation 'groups' not found in obs columns.\n",
      "\n",
      "Available columns:\n",
      " ['sample', 'treatment', 'cell_line', 'final_annotation', 'final_annotation_functional', 'nn_majority_label', 'nn_majority_frac', 'flag_misplaced']\n",
      "Please manually provide an appropriate annotation file to SimiCPipeline in SimiCExampleRun/inputFiles\n",
      "\n",
      "-------\n",
      "\n",
      "Experiment files saved successfully.\n",
      "\n",
      "-------\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# This takes 1.5 mins\n",
    "experiment.save_experiment_files(\n",
    "    run_data = subset_data,\n",
    "    matrix_filename = 'expression_matrix.csv',\n",
    "    tf_filename = 'TF_list.csv',\n",
    "    annotation = 'groups' # Will raise warning if column not found in subset_data.obs_names\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved expression matrix to SimiCExampleRun/inputFiles/expression_matrix.pickle\n",
      "\n",
      "Warning: Output file SimiCExampleRun/inputFiles/TF_list.csv already exists and will be overwritten.\n",
      "Saved 100 TFs to SimiCExampleRun/inputFiles/TF_list.csv\n",
      "\n",
      "-------\n",
      "\n",
      "Annotation 'treatment' found in obs columns!\n",
      "Warning: annotation is not numeric. Will convert from categorical to numeric.\n",
      "\n",
      "Annotation distribution: {0: 17259, 1: 20491, 2: 15426, 3: 19474}\n",
      "Saved annotation to SimiCExampleRun/inputFiles/treatment_annotation.txt\n",
      "\n",
      "-------\n",
      "\n",
      "Experiment files saved successfully.\n",
      "\n",
      "-------\n",
      "\n"
     ]
    }
   ],
   "source": [
    "experiment.save_experiment_files(\n",
    "    run_data = subset_data,\n",
    "    matrix_filename = 'expression_matrix.pickle',\n",
    "    tf_filename = 'TF_list.csv', # Will raise warning if file already exists and overwrite\n",
    "    annotation = 'treatment' \n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "<b>Success!</b> All preprocessing steps completed. Your files are ready for SimiC analysis.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Step 2.6: Verify Saved Files\n",
    "\n",
    "Check that all files were created correctly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SimiCExampleRun/\n",
      "├── inputFiles/\n",
      "│   ├── TF_list.csv\n",
      "│   ├── expression_matrix.csv\n",
      "│   ├── expression_matrix.pickle\n",
      "│   └── treatment_annotation.txt\n",
      "├── magic_output/\n",
      "│   ├── magic_imputed.h5ad\n",
      "│   └── magic_imputed.pickle\n",
      "└── outputSimic/\n",
      "    ├── figures/\n",
      "    └── matrices/\n"
     ]
    }
   ],
   "source": [
    "experiment.print_project_info(max_depth=2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Summary\n",
    "\n",
    "This tutorial covered:\n",
    "\n",
    "✓ Loading and filtering scRNA-seq data\n",
    "\n",
    "✓ Running MAGIC imputation\n",
    "\n",
    "✓ Selecting top variable genes using MAD\n",
    "\n",
    "✓ Preparing input files for SimiC analysis with proper directory structure\n",
    "\n",
    "### Output Directory Structure\n",
    "\n",
    "Your output directory now contains:\n",
    "```\n",
    "SimicExampleRun/\n",
    "├── magic_output/\n",
    "│   └── magic_imputed.pickle\n",
    "├── inputFiles/\n",
    "│   ├── expression_matrix.pickle\n",
    "│   ├── TF_list.pickle\n",
    "│   └── groups_phenotype.txt\n",
    "└── outputSimic/\n",
    "    ├── figures/\n",
    "    └── matrices/\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "# Alternative Approach\n",
    "\n",
    "In this the previous section we used the whole Magic-inputed matrix (obtained in [Part1](#part1)) and selected top MAD genes but you may want to run SimiC in a subset of cells from your data.\n",
    "\n",
    "Generally we recommend to impute the data in the whole dataset, especially if it was generated in the same sequencing batch, as MAGIC will have more context information to impute the data. However, we acknowledge that every experiment/dataset is different and may require a different approach.\n",
    "\n",
    "If you want to run SimiCPipeline in a subset of cells, once you have imputed your data, make sure you slice the adata object before you inilitalize the `ExperimentSetup` class so MAD genes are calculated over your cells of interest.\n",
    "\n",
    "Following this tutorial steps are not required for running `SimiCPipeline` but recommended before as it will facilitate the process. Just make sure that:\n",
    "\n",
    "1. You have **cell assignments** if not done before: Create a file with cell phenotype labels matching  the same order as your expression matrix that you will use in SimiC.\n",
    "2. Prepare **expression matrix** pickle/csv format\n",
    "3. Prepare **TF_list** pickle/csv with TF genes in your expression matrix.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will show how to easily subset your cells and save it with the `ExperimentSetup`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>treatment</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>final_annotation</th>\n",
       "      <th>final_annotation_functional</th>\n",
       "      <th>nn_majority_label</th>\n",
       "      <th>nn_majority_frac</th>\n",
       "      <th>flag_misplaced</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>43_01_73__s1</th>\n",
       "      <td>KPB25L-UV_Combination</td>\n",
       "      <td>Combination</td>\n",
       "      <td>KPB25L-UV</td>\n",
       "      <td>Cancer cells</td>\n",
       "      <td>Basal-like</td>\n",
       "      <td>Basal-like</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43_01_92__s1</th>\n",
       "      <td>KPB25L-UV_Combination</td>\n",
       "      <td>Combination</td>\n",
       "      <td>KPB25L-UV</td>\n",
       "      <td>Cancer cells</td>\n",
       "      <td>Proliferating cells</td>\n",
       "      <td>Proliferating cells</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43_01_94__s1</th>\n",
       "      <td>KPB25L-UV_Combination</td>\n",
       "      <td>Combination</td>\n",
       "      <td>KPB25L-UV</td>\n",
       "      <td>Cancer cells</td>\n",
       "      <td>Proliferating cells</td>\n",
       "      <td>Proliferating cells</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43_02_41__s1</th>\n",
       "      <td>KPB25L-UV_Combination</td>\n",
       "      <td>Combination</td>\n",
       "      <td>KPB25L-UV</td>\n",
       "      <td>Endothelial cells</td>\n",
       "      <td>Endothelial cells</td>\n",
       "      <td>Endothelial cells</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43_02_56__s1</th>\n",
       "      <td>KPB25L-UV_Combination</td>\n",
       "      <td>Combination</td>\n",
       "      <td>KPB25L-UV</td>\n",
       "      <td>Cancer cells</td>\n",
       "      <td>Basal-like</td>\n",
       "      <td>Basal-like</td>\n",
       "      <td>0.9</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>06_92_89__s8</th>\n",
       "      <td>KPB25L_control</td>\n",
       "      <td>control</td>\n",
       "      <td>KPB25L</td>\n",
       "      <td>Cancer cells</td>\n",
       "      <td>Basal-like</td>\n",
       "      <td>Basal-like</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>06_94_61__s8</th>\n",
       "      <td>KPB25L_control</td>\n",
       "      <td>control</td>\n",
       "      <td>KPB25L</td>\n",
       "      <td>Macrophages</td>\n",
       "      <td>Macrophages</td>\n",
       "      <td>Macrophages</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>06_96_48__s8</th>\n",
       "      <td>KPB25L_control</td>\n",
       "      <td>control</td>\n",
       "      <td>KPB25L</td>\n",
       "      <td>Cancer cells</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>06_96_58__s8</th>\n",
       "      <td>KPB25L_control</td>\n",
       "      <td>control</td>\n",
       "      <td>KPB25L</td>\n",
       "      <td>Macrophages</td>\n",
       "      <td>Macrophages</td>\n",
       "      <td>Macrophages</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>06_96_82__s8</th>\n",
       "      <td>KPB25L_control</td>\n",
       "      <td>control</td>\n",
       "      <td>KPB25L</td>\n",
       "      <td>Cancer cells</td>\n",
       "      <td>Basal-like</td>\n",
       "      <td>Basal-like</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>72650 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             sample    treatment  cell_line  \\\n",
       "Cell                                                          \n",
       "43_01_73__s1  KPB25L-UV_Combination  Combination  KPB25L-UV   \n",
       "43_01_92__s1  KPB25L-UV_Combination  Combination  KPB25L-UV   \n",
       "43_01_94__s1  KPB25L-UV_Combination  Combination  KPB25L-UV   \n",
       "43_02_41__s1  KPB25L-UV_Combination  Combination  KPB25L-UV   \n",
       "43_02_56__s1  KPB25L-UV_Combination  Combination  KPB25L-UV   \n",
       "...                             ...          ...        ...   \n",
       "06_92_89__s8         KPB25L_control      control     KPB25L   \n",
       "06_94_61__s8         KPB25L_control      control     KPB25L   \n",
       "06_96_48__s8         KPB25L_control      control     KPB25L   \n",
       "06_96_58__s8         KPB25L_control      control     KPB25L   \n",
       "06_96_82__s8         KPB25L_control      control     KPB25L   \n",
       "\n",
       "               final_annotation final_annotation_functional  \\\n",
       "Cell                                                          \n",
       "43_01_73__s1       Cancer cells                  Basal-like   \n",
       "43_01_92__s1       Cancer cells         Proliferating cells   \n",
       "43_01_94__s1       Cancer cells         Proliferating cells   \n",
       "43_02_41__s1  Endothelial cells           Endothelial cells   \n",
       "43_02_56__s1       Cancer cells                  Basal-like   \n",
       "...                         ...                         ...   \n",
       "06_92_89__s8       Cancer cells                  Basal-like   \n",
       "06_94_61__s8        Macrophages                 Macrophages   \n",
       "06_96_48__s8       Cancer cells                     Unknown   \n",
       "06_96_58__s8        Macrophages                 Macrophages   \n",
       "06_96_82__s8       Cancer cells                  Basal-like   \n",
       "\n",
       "                nn_majority_label  nn_majority_frac  flag_misplaced  \n",
       "Cell                                                                 \n",
       "43_01_73__s1           Basal-like               1.0           False  \n",
       "43_01_92__s1  Proliferating cells               1.0           False  \n",
       "43_01_94__s1  Proliferating cells               1.0           False  \n",
       "43_02_41__s1    Endothelial cells               1.0           False  \n",
       "43_02_56__s1           Basal-like               0.9           False  \n",
       "...                           ...               ...             ...  \n",
       "06_92_89__s8           Basal-like               1.0           False  \n",
       "06_94_61__s8          Macrophages               1.0           False  \n",
       "06_96_48__s8              Unknown               1.0           False  \n",
       "06_96_58__s8          Macrophages               1.0           False  \n",
       "06_96_82__s8           Basal-like               1.0           False  \n",
       "\n",
       "[72650 rows x 8 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imputed_data.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of selected cells: {21490}\n",
      "Suset matrix shape: {(21490, 27837)}\n",
      "Creating project directory: SimiCExampleRun/KPB25L/Tumor\n",
      "Tumor/\n",
      "├── inputFiles/\n",
      "└── outputSimic/\n"
     ]
    }
   ],
   "source": [
    "from simicpipeline import ExperimentSetup\n",
    "cell_mask = imputed_data.obs['cell_line'].isin([\"KPB25L\"]) & imputed_data.obs['final_annotation_functional'].isin(['Proliferating cells','Basal-like'])\n",
    "print(f\"Number of selected cells:\",{sum(cell_mask)})\n",
    "subset_imputed_data = imputed_data[cell_mask,:].copy()\n",
    "\n",
    "print(f\"Suset matrix shape:\", {subset_imputed_data.shape})\n",
    "\n",
    "experiment2 = ExperimentSetup(\n",
    "    input_data = subset_imputed_data, \n",
    "    tf_path = \"./data/TF_list.csv\", # Should have no header\n",
    "    project_dir='./SimiCExampleRun/KPB25L/Tumor'\n",
    ")\n",
    "experiment2.print_project_info(max_depth=1)\n",
    "# Then follow the same steps as above to select genes and save experiment files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "However the initial output directory will then look like:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SimiCExampleRun/\n",
      "├── KPB25L/\n",
      "│   └── Tumor/\n",
      "│       ├── inputFiles/\n",
      "│       └── outputSimic/\n",
      "│           ├── figures/\n",
      "│           └── matrices/\n",
      "├── inputFiles/\n",
      "│   ├── TF_list.csv\n",
      "│   ├── expression_matrix.csv\n",
      "│   ├── expression_matrix.pickle\n",
      "│   └── treatment_annotation.txt\n",
      "├── magic_output/\n",
      "│   ├── magic_imputed.h5ad\n",
      "│   └── magic_imputed.pickle\n",
      "└── outputSimic/\n",
      "    ├── figures/\n",
      "    └── matrices/\n",
      "        └── full_experiment/\n"
     ]
    }
   ],
   "source": [
    "experiment.print_project_info(max_depth=4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Repeat the steps to calculate MAD genes adn save files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Removing 8 targets with MAD = 0\n",
      "Selecting top 1000 targets based on MAD.\n",
      "Saved expression matrix to SimiCExampleRun/KPB25L/Tumor/inputFiles/expression_matrix.pickle\n",
      "Saved 100 TFs to SimiCExampleRun/KPB25L/Tumor/inputFiles/TF_list.csv\n",
      "\n",
      "-------\n",
      "\n",
      "Annotation 'treatment' found in obs columns!\n",
      "Warning: annotation is not numeric. Will convert from categorical to numeric.\n",
      "\n",
      "Annotation distribution: {0: 4392, 1: 5895, 2: 4964, 3: 6239}\n",
      "Saved annotation to SimiCExampleRun/KPB25L/Tumor/inputFiles/treatment_annotation.txt\n",
      "\n",
      "-------\n",
      "\n",
      "Experiment files saved successfully.\n",
      "\n",
      "-------\n",
      "\n",
      "Subset data shape: (21490, 1100)\n"
     ]
    }
   ],
   "source": [
    "tf_list, target_list = experiment2.calculate_mad_genes(\n",
    "    n_tfs=100,\n",
    "    n_targets=1000\n",
    ")\n",
    "# Combine TF and target lists\n",
    "selected_genes2 = tf_list + target_list\n",
    "subset_experiment_data = subset_imputed_data[:, selected_genes2].copy()\n",
    "\n",
    "experiment2.save_experiment_files(\n",
    "    run_data = subset_experiment_data,\n",
    "    matrix_filename = 'expression_matrix.pickle',\n",
    "    tf_filename = 'TF_list.csv', # Will raise warning if file already exists and overwrite\n",
    "    annotation = 'treatment' \n",
    ")\n",
    "print(f\"Subset data shape: {subset_experiment_data.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tumor/\n",
      "├── inputFiles/\n",
      "│   ├── TF_list.csv\n",
      "│   ├── expression_matrix.pickle\n",
      "│   └── treatment_annotation.txt\n",
      "└── outputSimic/\n",
      "    ├── figures/\n",
      "    └── matrices/\n"
     ]
    }
   ],
   "source": [
    "experiment2.print_project_info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "# Next steps:\n",
    "\n",
    "1. **Run SimiC**: Use `SimiCPipeline` class to run SimiC.\n",
    "2. **Explore results**: Use `SimicVisualization` class to analyze GRNs and TF activities.\n",
    "\n",
    "Check `Tutorial_SimiCPipeline_full.ipynb` or `Tutorial_SimiCPipeline_visualization` for guided info.\n",
    "\n",
    "\n",
    "### Final Notes\n",
    "\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "<b>Data Format:</b> All matrices are stored as cells × genes (rows = cells, columns = genes)\n",
    "</div>\n",
    "\n",
    "<div class=\"alert alert-block alert-warning\">\n",
    "<b>Memory Usage:</b> MAGIC imputation can be memory-intensive for large datasets. Consider using a machine with sufficient RAM and adjusting MAGIC parameters (n_jobs, knn, t)\n",
    "</div>\n",
    "\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "<b>Please note:</b> Although you will be able to pass custom file/direcotry paths, we highly recommend to follow the directory structure described above and follow this tutorial before running SimiC to avoid errors.\n",
    "</div>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SimiC Pipeline Kernel",
   "language": "python",
   "name": "simicpipeline"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
